integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala

```diff
@@ -0,0 +1,103 @@
+import org.apache.carbondata.core.indexstore.BlockletDataMapIndexWrapper
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable
+import org.apache.carbondata.datamap.bloom.BloomCacheKeyValue
+import org.apache.carbondata.events.{DropCacheEvent, OperationContext, OperationListenerBus}
+
+case class CarbonDropCacheCommand(tableIdentifier: TableIdentifier, internalCall: Boolean)
+  extends MetadataCommand {
+
+  override def processMetadata(sparkSession: SparkSession): Seq[Row] = {
+    val carbonTable = CarbonEnv.getCarbonTable(tableIdentifier)(sparkSession)
+    if (carbonTable.isChildDataMap && !internalCall) {
+      throw new UnsupportedOperationException("Operation not allowed on child table.")
+    }
+    clearCache(carbonTable, sparkSession)
+    Seq.empty
+  }
+
+  def clearCache(carbonTable: CarbonTable, sparkSession: SparkSession): Unit = {
+    val cache = CacheProvider.getInstance().getCarbonCache
+    if (cache != null) {
+      val dbLocation = CarbonEnv
+        .getDatabaseLocation(tableIdentifier.database
+          .getOrElse(sparkSession.catalog.currentDatabase), sparkSession)
+        .replace(CarbonCommonConstants.WINDOWS_FILE_SEPARATOR,
+          CarbonCommonConstants.FILE_SEPARATOR)
+      val tablePath = carbonTable.getTablePath + CarbonCommonConstants.FILE_SEPARATOR
+
+      // Dictionary IDs
+      val dictIds = carbonTable.getAllDimensions.asScala.filter(_.isGlobalDictionaryEncoding)
+        .map(_.getColumnId).toArray
+
+      // Remove elements from cache
+      val keysToRemove = ListBuffer[String]()
+      val cacheIterator = cache.getCacheMap.entrySet().iterator()
+      while (cacheIterator.hasNext) {
+        val entry = cacheIterator.next()
+        val cache = entry.getValue
+
+        if (cache.isInstanceOf[BlockletDataMapIndexWrapper]) {
+          // index
+          val indexPath = entry.getKey.replace(CarbonCommonConstants.WINDOWS_FILE_SEPARATOR,
+            CarbonCommonConstants.FILE_SEPARATOR)
+          if (indexPath.startsWith(tablePath)) {
+            keysToRemove += entry.getKey
+          }
+        } else if (cache.isInstanceOf[BloomCacheKeyValue.CacheValue]) {
+          // bloom datamap
+          val shardPath = entry.getKey.replace(CarbonCommonConstants.WINDOWS_FILE_SEPARATOR,
+            CarbonCommonConstants.FILE_SEPARATOR)
+          if (shardPath.contains(tablePath)) {
+            keysToRemove += entry.getKey
+          }
+        } else if (cache.isInstanceOf[AbstractColumnDictionaryInfo]) {
+          // dictionary
+          val dictId = dictIds.find(id => entry.getKey.startsWith(id))
+          if (dictId.isDefined) {
+            keysToRemove += entry.getKey
+          }
+        }
+      }
+      cache.removeAll(keysToRemove.asJava)
+
+      val dropCacheEvent = DropCacheEvent(
+        carbonTable,
+        sparkSession
+      )
+      val operationContext = new OperationContext
+      OperationListenerBus.getInstance.fireEvent(dropCacheEvent, operationContext)
```

Comment:
Event should be fired before main table cache invalidation.

Reply:
Done.